src/build/env.py

```diff
@@ -1,312 +1,313 @@
+    return executable, script_dir
 
-        The default implementation uses the logging module but this function can be
-        overwritten by users to have a different implementation.
 
-        :param msg: Message to output
-        """
-        if sys.version_info >= (3, 8):
-            _logger.log(logging.INFO, message, stacklevel=2)
-        else:
-            _logger.log(logging.INFO, message)
+@_cache()
+def _should_use_virtualenv() -> bool:
+    import packaging.requirements
 
+    # virtualenv might be incompatible if it was installed separately
+    # from build. This verifies that virtualenv and all of its
+    # dependencies are installed as specified by build.
+    return not any(
+        packaging.requirements.Requirement(u).name == 'virtualenv' for d in check_dependency('build[virtualenv]') for u in d
+    )
 
-class _IsolatedEnvVenvPip(IsolatedEnv):
-    """
-    Isolated build environment context manager
 
-    Non-standard paths injected directly to sys.path will still be passed to the environment.
-    """
+@_cache()
+def _get_min_pip_version() -> str:
+    if platform.system() == 'Darwin':
+        release, _, machine = platform.mac_ver()
 
-    def __init__(
-        self,
-        path: str,
-        python_executable: str,
-        scripts_dir: str,
-        log: Callable[[str], None],
-    ) -> None:
-        """
-        :param path: The path where the environment exists
-        :param python_executable: The python executable within the environment
-        :param log: Log function
-        """
-        self._path = path
-        self._python_executable = python_executable
-        self._scripts_dir = scripts_dir
-        self._log = log
+        # Apple Silicon support for wheels shipped with packaging 20.9,
+        # vendored in pip 21.0.1.
+        if machine == 'arm64':
+            return '21.0.1'
 
-    @property
-    def path(self) -> str:
-        """The location of the isolated build environment."""
-        return self._path
+        # macOS 11+ name scheme change requires 20.3. Intel macOS 11.0 can
+        # be told to report 10.16 for backwards compatibility;
+        # but that also fixes earlier versions of pip so this is only needed for 11+.
+        major_version = int(release[: release.find('.')])
+        if major_version >= 11:
+            return '20.3'
 
-    @property
-    def executable(self) -> str:
-        """The python executable of the isolated build environment."""
-        return self._python_executable
+    # PEP 517 and manylinux1 were first implemented in 19.1
+    return '19.1'
 
-    @property
-    def scripts_dir(self) -> str:
-        return self._scripts_dir
 
-    def install(self, requirements: Iterable[str]) -> None:
-        """
-        Install packages from PEP 508 requirements in the isolated build environment.
+class _DefaultIsolatedEnv(IsolatedEnv):
+    """
+    An isolated environment class which combines venv and virtualenv with pip.
```

Comment:
I think you wanted to say here `or` and not `and`. Only one can be active at the same time.

Reply:
Done.